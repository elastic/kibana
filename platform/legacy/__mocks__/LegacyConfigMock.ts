/**
 * This is a partial mock of src/server/config/config.js.
 */
export class LegacyConfigMock {
  constructor(public __rawData: Map<string, any> = new Map()) {}

  readonly set = jest.fn((key, value) => {
    // Real legacy config throws error if key is not presented in the schema.
    if (!this.__rawData.has(key)) {
      throw new TypeError(`Unknown schema key: ${key}`);
    }

    this.__rawData.set(key, value);
  });

  readonly get = jest.fn(key => {
    // Real legacy config throws error if key is not presented in the schema.
    if (!this.__rawData.has(key)) {
      throw new TypeError(`Unknown schema key: ${key}`);
    }

    return this.__rawData.get(key);
  });

  readonly has = jest.fn(key => this.__rawData.has(key));
}
