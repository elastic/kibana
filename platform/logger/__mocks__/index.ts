// Test helpers to simplify mocking logs and collecting all their outputs

const _log = {
  debug: jest.fn(),
  info: jest.fn(),
  error: jest.fn()
};

const _clear = () => {
  logger.get.mockClear();
  _log.debug.mockClear();
  _log.info.mockClear();
  _log.error.mockClear();
}

const _collect = () => ({
  debug: _log.debug.mock.calls,
  info: _log.info.mock.calls,
  error: _log.error.mock.calls
});

export const logger = {
  get: jest.fn(() => _log),
  _log,
  _collect,
  _clear
};