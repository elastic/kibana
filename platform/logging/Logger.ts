import { Appender } from './appenders/Appenders';
import { LogLevel } from './LogLevel';
import { LogRecord } from './LogRecord';

/**
 * Logger exposes all the necessary methods to log any type of information and
 * this is the interface used by the logging consumers including plugins.
 */
export interface Logger {
  trace(message: string, meta?: { [key: string]: any }): void;
  debug(message: string, meta?: { [key: string]: any }): void;
  info(message: string, meta?: { [key: string]: any }): void;
  warn(errorOrMessage: string | Error, meta?: { [key: string]: any }): void;
  error(errorOrMessage: string | Error, meta?: { [key: string]: any }): void;
  fatal(errorOrMessage: string | Error, meta?: { [key: string]: any }): void;

  /** @internal */
  log(record: LogRecord): void;
}

/** @internal */
export class BaseLogger implements Logger {
  constructor(
    private readonly context: string,
    private readonly level: LogLevel,
    private readonly appenders: Appender[]
  ) {}

  trace(message: string, meta?: { [key: string]: any }): void {
    this.log(this.createLogRecord(LogLevel.Trace, message, meta));
  }

  debug(message: string, meta?: { [key: string]: any }): void {
    this.log(this.createLogRecord(LogLevel.Debug, message, meta));
  }

  info(message: string, meta?: { [key: string]: any }): void {
    this.log(this.createLogRecord(LogLevel.Info, message, meta));
  }

  warn(errorOrMessage: string | Error, meta?: { [key: string]: any }): void {
    this.log(this.createLogRecord(LogLevel.Warn, errorOrMessage, meta));
  }

  error(errorOrMessage: string | Error, meta?: { [key: string]: any }): void {
    this.log(this.createLogRecord(LogLevel.Error, errorOrMessage, meta));
  }

  fatal(errorOrMessage: string | Error, meta?: { [key: string]: any }): void {
    this.log(this.createLogRecord(LogLevel.Fatal, errorOrMessage, meta));
  }

  log(record: LogRecord) {
    if (!this.level.supports(record.level)) {
      return;
    }

    for (const appender of this.appenders) {
      appender.append(record);
    }
  }

  private createLogRecord(
    level: LogLevel,
    errorOrMessage: string | Error,
    meta?: { [key: string]: any }
  ): LogRecord {
    const isError = errorOrMessage instanceof Error;
    return {
      timestamp: new Date(),
      level,
      context: this.context,
      message: isError
        ? (errorOrMessage as Error).message
        : errorOrMessage as string,
      error: isError ? errorOrMessage as Error : undefined,
      meta
    };
  }
}

/** @internal */
export class LoggerAdapter implements Logger {
  constructor(public logger: Logger) {}

  trace(message: string, meta?: { [key: string]: any }): void {
    this.logger.trace(message, meta);
  }

  debug(message: string, meta?: { [key: string]: any }): void {
    this.logger.debug(message, meta);
  }

  info(message: string, meta?: { [key: string]: any }): void {
    this.logger.info(message, meta);
  }

  warn(errorOrMessage: string | Error, meta?: { [key: string]: any }): void {
    this.logger.warn(errorOrMessage, meta);
  }

  error(errorOrMessage: string | Error, meta?: { [key: string]: any }): void {
    this.logger.error(errorOrMessage, meta);
  }

  fatal(errorOrMessage: string | Error, meta?: { [key: string]: any }): void {
    this.logger.fatal(errorOrMessage, meta);
  }

  log(record: LogRecord) {
    this.logger.log(record);
  }
}
