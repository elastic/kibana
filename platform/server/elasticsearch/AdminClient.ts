import { Client } from 'elasticsearch';

export class AdminClient {
  constructor(
    private readonly client: Client,
  ) {
  }

  call(endpoint: string, clientParams = {}, options = {}): any {
    this.client.search({});
    return {};
  }
}

