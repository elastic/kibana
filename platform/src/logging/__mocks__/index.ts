// Test helpers to simplify mocking logs and collecting all their outputs

const _log = {
  debug: jest.fn(),
  info: jest.fn(),
  warn: jest.fn(),
  error: jest.fn(),
  trace: jest.fn(),
  fatal: jest.fn(),
  log: jest.fn(),
};

const _clear = () => {
  logger.get.mockClear();
  _log.debug.mockClear();
  _log.info.mockClear();
  _log.warn.mockClear();
  _log.error.mockClear();
  _log.trace.mockClear();
  _log.fatal.mockClear();
  _log.log.mockClear();
};

const _collect = () => ({
  debug: _log.debug.mock.calls,
  info: _log.info.mock.calls,
  warn: _log.warn.mock.calls,
  error: _log.error.mock.calls,
  trace: _log.trace.mock.calls,
  fatal: _log.fatal.mock.calls,
  log: _log.log.mock.calls,
});

export const logger = {
  get: jest.fn(() => _log),
  _log,
  _collect,
  _clear,
};
