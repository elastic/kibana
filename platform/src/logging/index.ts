export { Logger } from './Logger';
export { LoggerFactory } from './LoggerFactory';
