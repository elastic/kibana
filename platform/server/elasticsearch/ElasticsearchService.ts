import {
  Observable,
  Subscription,
  k$,
  map,
  filter,
  switchMap,
  shareLast
} from 'kbn-observable';

import { ElasticsearchClusterType } from './ElasticsearchConfig';
import { ElasticsearchConfigs } from './ElasticsearchConfigs';
import { Cluster } from './Cluster';
import { LoggerFactory } from '../../logging';
import { CoreService } from '../../types/CoreService';

type Clusters = { [type in ElasticsearchClusterType]: Cluster };

export class ElasticsearchService implements CoreService {
  private clusters$: Observable<Clusters>;
  private subscription?: Subscription;

  constructor(
    config$: Observable<ElasticsearchConfigs>,
    logger: LoggerFactory
  ) {
    const log = logger.get('elasticsearch');

    this.clusters$ = k$(config$)(
      filter(() => {
        if (this.subscription !== undefined) {
          log.error('clusters cannot be changed after they are created');
          return false;
        }

        return true;
      }),
      switchMap(
        configs =>
          new Observable<Clusters>(observer => {
            log.info('creating Elasticsearch clusters');

            const clusters = {
              data: new Cluster(configs.forType('data'), logger),
              admin: new Cluster(configs.forType('admin'), logger)
            };

            observer.next(clusters);

            return () => {
              log.info('closing Elasticsearch clusters');

              clusters.data.close();
              clusters.admin.close();
            };
          })
      ),
      // We only want a single subscription of this as we only want to create a
      // single set of clusters at a time. We therefore share these, plus we
      // always replay the latest set of clusters when subscribing.
      shareLast()
    );
  }

  async start() {
    // ensure that we don't unnecessarily re-create clusters by always having
    // at least one current connection
    this.subscription = this.clusters$.subscribe();
  }

  async stop() {
    if (this.subscription !== undefined) {
      this.subscription.unsubscribe();
    }
  }

  getClusterOfType$(type: ElasticsearchClusterType) {
    return k$(this.clusters$)(map(clusters => clusters[type]));
  }
}
