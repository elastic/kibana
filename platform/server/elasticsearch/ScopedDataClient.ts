import { Client } from 'elasticsearch';
import { Headers } from '../http/Router/headers';

//export class ScopedDataClient extends DataClient {
export class ScopedDataClient {
  constructor(
    private readonly client: Client,
    headers: Headers,
  ) {
  }

  call(endpoint: string, clientParams = {}, options = {}): any {
    // TODO: use this.headers
    return this.client.search({});
  }
}
