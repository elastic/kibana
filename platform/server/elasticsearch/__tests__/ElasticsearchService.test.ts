const mockCreateClient = jest.fn();

jest.mock('elasticsearch', () => ({
  Client: mockCreateClient,
}));

import { ElasticsearchService } from '../ElasticsearchService';
import { AdminClient } from '../AdminClient';
import { ScopedDataClient } from '../ScopedDataClient';
import { logger } from '../../../logging/__mocks__';
import {
  k$,
  BehaviorSubject,
  first,
  toPromise,
} from '@elastic/kbn-observable';

test('should not create multiple clients', async () => {
  const forTypeContents = {
    toElasticsearchClientConfig: () => {},
    requestHeadersWhitelist: [],
  };

  const elasticsearchConfigs = {
    forType: () => forTypeContents,
  };

  const configs$: any = new BehaviorSubject(elasticsearchConfigs);

  const service = new ElasticsearchService(configs$, logger);

  mockCreateClient.mockReset();

  service.start();

  await service.getScopedDataClient({ foo: 'bar' });
  await service.getScopedDataClient({ foo: 'bar' });

  service.stop();

  expect(mockCreateClient).toHaveBeenCalledTimes(2);
});

test('should get an AdminClient', async () => {
  const elasticsearchConfigs = {
    forType: () => ({ toElasticsearchClientConfig: () => {} }),
  };

  const configs$: any = new BehaviorSubject(elasticsearchConfigs);

  const service = new ElasticsearchService(configs$, logger);
  const adminClient = await k$(service.getAdminClient$())(first(), toPromise());

  expect(adminClient).toBeInstanceOf(AdminClient);
});

test('should get a ScopedDataClient', async () => {
  const forTypeContents = {
    toElasticsearchClientConfig: () => {},
    requestHeadersWhitelist: [],
  };

  const elasticsearchConfigs = {
    forType: () => forTypeContents,
  };

  const configs$: any = new BehaviorSubject(elasticsearchConfigs);

  const service = new ElasticsearchService(configs$, logger);
  const dataClient = await service.getScopedDataClient({ foo: 'bar' });

  expect(dataClient).toBeInstanceOf(ScopedDataClient);
});

test('should get a ScopedDataClient observable', async () => {
  const forTypeContents = {
    toElasticsearchClientConfig: () => {},
    requestHeadersWhitelist: [],
  };

  const elasticsearchConfigs = {
    forType: () => forTypeContents,
  };

  const configs$: any = new BehaviorSubject(elasticsearchConfigs);

  const service = new ElasticsearchService(configs$, logger);
  const dataClient$ = service.getScopedDataClient$({ foo: 'bar' });

  expect(await k$(dataClient$)(first(), toPromise())).toBeInstanceOf(ScopedDataClient);
});
