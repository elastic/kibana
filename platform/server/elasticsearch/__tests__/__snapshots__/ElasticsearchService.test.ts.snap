// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not create multiple clients while service is running 1`] = `
Array [
  Array [
    Object {
      "options": undefined,
      "type": "data",
    },
  ],
  Array [
    Object {
      "options": Object {
        "shouldAuth": false,
      },
      "type": "admin",
    },
  ],
]
`;
